File: TypeScript/sd-def-schema.ts
// sd-def-schema.ts (generated)

export interface Asset {
  id: string;
  name: string;
  type: "host" | "service" | "application" | "user" | "network";
  tags?: string[];
  criticality?: "low" | "medium" | "high" | "critical";
}

export interface ControlFrameworkRef {
  framework: "NIST-CSF";
  function: "Identify" | "Protect" | "Detect" | "Respond" | "Recover";
  category: string;
  subcategory: string;
}

export interface Control {
  id: string;
  name: string;
  frameworkRef: ControlFrameworkRef;
  owner?: string;
}

export interface SignalBase {
  id: string;
  source: string;
  timestamp: string; // ISO
  assetId?: string;
  controlId?: string;
  severity?: "info" | "low" | "medium" | "high" | "critical";
  details?: Record<string, unknown>;
}

export type DegradationEventType =
  | "misconfiguration"
  | "process_drift"
  | "visibility_loss"
  | "control_failure"
  | "surface_expansion";

export interface DegradationEvent {
  id: string;
  type: DegradationEventType;
  assetId: string;
  controlId?: string;
  timestamp: string;
  signalIds: string[];
  impactSurfaceId?: string;
}

export interface ImpactSurface {
  id: string;
  assetIds: string[];
  userSegments?: string[];
  dataClasses?: string[];
}

export interface ScoreBase {
  id: string;
  eventId: string;
  value: number;
  scale: number;
  dimension?: string;
}

export interface ControlWeakeningScore extends ScoreBase {
  dimension: "control_weakening";
}

export interface ExposureAmplificationScore extends ScoreBase {
  dimension: "exposure_amplification";
}

export interface DetectionBlindnessScore extends ScoreBase {
  dimension: "detection_blindness";
}

export interface ProcessErosionScore extends ScoreBase {
  dimension: "process_erosion";
}

export type AnyScore =
  | ControlWeakeningScore
  | ExposureAmplificationScore
  | DetectionBlindnessScore
  | ProcessErosionScore;

export type SDIScopeType = "asset" | "control" | "global";

export interface SDIScope {
  type: SDIScopeType;
  assetId?: string;
  controlId?: string;
}

export interface SecurityDegradationIndex {
  id: string;
  scope: SDIScope;
  value: number;
  scale: number;
  components?: {
    controlWeakening?: number;
    exposureAmplification?: number;
    detectionBlindness?: number;
    processErosion?: number;
  };
}

export type RemediationType =
  | "immediate_fix"
  | "structural_remediation"
  | "control_hardening"
  | "visibility_restoration";

export type RemediationPriority = "p0" | "p1" | "p2" | "p3";

export interface RemediationAction {
  id: string;
  type: RemediationType;
  priority: RemediationPriority;
  title: string;
  description?: string;
  relatedEventIds?: string[];
  owner?: string;
}

File: python/sd_def_schema.py
# sd_def_schema.py (generated)

from typing import List, Optional, Literal, Dict, Any
from pydantic import BaseModel


class Asset(BaseModel):
    id: str
    name: str
    type: Literal["host", "service", "application", "user", "network"]
    tags: Optional[List[str]] = None
    criticality: Optional[Literal["low", "medium", "high", "critical"]] = None


class ControlFrameworkRef(BaseModel):
    framework: Literal["NIST-CSF"]
    function: Literal["Identify", "Protect", "Detect", "Respond", "Recover"]
    category: str
    subcategory: str


class Control(BaseModel):
    id: str
    name: str
    frameworkRef: ControlFrameworkRef
    owner: Optional[str] = None


class SignalBase(BaseModel):
    id: str
    source: str
    timestamp: str  # ISO 8601
    assetId: Optional[str] = None
    controlId: Optional[str] = None
    severity: Optional[Literal["info", "low", "medium", "high", "critical"]] = None
    details: Optional[Dict[str, Any]] = None


DegradationEventType = Literal[
    "misconfiguration",
    "process_drift",
    "visibility_loss",
    "control_failure",
    "surface_expansion",
]


class DegradationEvent(BaseModel):
    id: str
    type: DegradationEventType
    assetId: str
    controlId: Optional[str] = None
    timestamp: str
    signalIds: List[str]
    impactSurfaceId: Optional[str] = None


class ImpactSurface(BaseModel):
    id: str
    assetIds: List[str]
    userSegments: Optional[List[str]] = None
    dataClasses: Optional[List[str]] = None


class ScoreBase(BaseModel):
    id: str
    eventId: str
    value: float
    scale: float
    dimension: Optional[str] = None


class ControlWeakeningScore(ScoreBase):
    dimension: Literal["control_weakening"] = "control_weakening"


class ExposureAmplificationScore(ScoreBase):
    dimension: Literal["exposure_amplification"] = "exposure_amplification"


class DetectionBlindnessScore(ScoreBase):
    dimension: Literal["detection_blindness"] = "detection_blindness"


class ProcessErosionScore(ScoreBase):
    dimension: Literal["process_erosion"] = "process_erosion"


SDIScopeType = Literal["asset", "control", "global"]


class SDIScope(BaseModel):
    type: SDIScopeType
    assetId: Optional[str] = None
    controlId: Optional[str] = None


class SecurityDegradationIndex(BaseModel):
    id: str
    scope: SDIScope
    value: float
    scale: float
    components: Optional[Dict[str, float]] = None


RemediationType = Literal[
    "immediate_fix",
    "structural_remediation",
    "control_hardening",
    "visibility_restoration",
]

RemediationPriority = Literal["p0", "p1", "p2", "p3"]


class RemediationAction(BaseModel):
    id: str
    type: RemediationType
    priority: RemediationPriority
    title: str
    description: Optional[str] = None
    relatedEventIds: Optional[List[str]] = None
    owner: Optional[str] = None
